File: web/src/api/questions.ts
import { apiClient } from './client';

/**
 * Questions API
 *
 * API functions for generating and submitting questions
 */

/**
 * Subject types for questions
 */
export type Subject = 'mathematics' | 'physics' | 'chemistry' | 'biology';

/**
 * Difficulty levels
 */
export type Difficulty = 'easy' | 'medium' | 'hard';

/**
 * Generated question from backend
 */
export interface Question {
  question: string;
  options: string[];
  correct_answer: string;
  explanation: string;
  subject: string;
  topic: string;
  subtopic?: string;
  difficulty: string;
  class_level: number;
  question_type: string;
  time_estimate_seconds: number;
  concept_tags: string[];
}

/**
 * Request to generate questions
 */
export interface GenerateQuestionsRequest {
  subject: string;
  topic: string;
  difficulty?: Difficulty;
  class_level: number;
  count?: number;
}

/**
 * Response from question generation
 */
export interface GenerateQuestionsResponse {
  questions: Question[];
  count: number;
  source: string;
}

/**
 * Question attempt submission
 */
export interface QuestionAttempt {
  question_id: string;
  selected_answer: string;
  is_correct: boolean;
  subject: string;
  topic: string;
  time_taken_seconds?: number;
}

/**
 * Topics available for a subject
 */
export interface SubjectTopics {
  subject: string;
  topics: string[];
}

/**
 * Generate questions using Gemini AI
 *
 * @param request - Question generation parameters
 * @returns Promise resolving to generated questions
 * @throws ApiError if request fails
 */
export const generateQuestions = async (
  request: GenerateQuestionsRequest
): Promise<GenerateQuestionsResponse> => {
  try {
    const response = await apiClient.post<GenerateQuestionsResponse>(
      '/questions/generate',
      request
    );
    return response.data;
  } catch (error) {
    console.error('Failed to generate questions:', error);
    throw error;
  }
};

/**
 * Submit a question attempt
 *
 * @param attempt - The attempt data to submit
 * @returns Promise resolving when submission is successful
 * @throws ApiError if submission fails
 */
export const submitQuestionAttempt = async (
  attempt: QuestionAttempt
): Promise<void> => {
  try {
    await apiClient.post('/questions/attempt', attempt);
  } catch (error) {
    console.error('Failed to submit question attempt:', error);
    throw error;
  }
};

/**
 * Submit multiple question attempts (batch)
 *
 * @param attempts - Array of attempts to submit
 * @returns Promise resolving when submission is successful
 * @throws ApiError if submission fails
 */
export const submitQuestionAttempts = async (
  attempts: QuestionAttempt[]
): Promise<void> => {
  try {
    await apiClient.post('/questions/attempts/batch', { attempts });
  } catch (error) {
    console.error('Failed to submit question attempts:', error);
    throw error;
  }
};

/**
 * Get available topics for a subject and class
 *
 * @param subject - The subject to get topics for
 * @param classLevel - The class level (10 or 12)
 * @returns Promise resolving to available topics
 * @throws ApiError if request fails
 */
export const getTopicsForSubject = async (
  subject: string,
  classLevel: number
): Promise<string[]> => {
  try {
    const response = await apiClient.get<{ topics: string[] }>('/questions/topics', {
      params: { subject, class_level: classLevel },
    });
    return response.data.topics;
  } catch (error) {
    console.error('Failed to get topics:', error);
    throw error;
  }
};

File: backend/app/api/v1/questions.py
"""
Question generation and management endpoints
"""
from fastapi import APIRouter, Depends, HTTPException, status
from typing import List

from app.core.security import get_current_user
from app.core.gemini import gemini_client
from app.schemas.question import (
    GenerateQuestionsRequest,
    GenerateQuestionsResponse,
    QuestionBase
)

router = APIRouter(prefix="/questions", tags=["questions"])


@router.post("/generate", response_model=GenerateQuestionsResponse)
async def generate_questions(
    request: GenerateQuestionsRequest,
    current_user: dict = Depends(get_current_user)
):
    """
    Generate questions dynamically using Google Gemini AI

    Requires authentication. Generates questions based on:
    - Subject (mathematics, science, etc.)
    - Topic within the subject
    - Difficulty level (easy, medium, hard)
    - Class level (10 or 12)
    - Number of questions (1-20)
    """
    try:
        # Validate class level
        if request.class_level not in [10, 12]:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Class level must be 10 or 12"
            )

        # Validate difficulty
        valid_difficulties = ["easy", "medium", "hard"]
        if request.difficulty.lower() not in valid_difficulties:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail=f"Difficulty must be one of: {', '.join(valid_difficulties)}"
            )

        # Generate questions using Gemini
        questions_data = await gemini_client.generate_questions(
            subject=request.subject,
            topic=request.topic,
            difficulty=request.difficulty,
            class_level=request.class_level,
            count=request.count
        )

        if not questions_data:
            raise HTTPException(
                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                detail="Failed to generate questions. Please try again."
            )

        # Convert to QuestionBase objects
        questions = []
        for q_data in questions_data:
            try:
                question = QuestionBase(
                    question=q_data.get("question", ""),
                    options=q_data.get("options", []),
                    correct_answer=q_data.get("correct_answer", ""),
                    explanation=q_data.get("explanation", ""),
                    subject=q_data.get("subject", request.subject),
                    topic=q_data.get("topic", request.topic),
                    difficulty=q_data.get("difficulty", request.difficulty),
                    class_level=request.class_level,
                    question_type="mcq"
                )
                questions.append(question)
            except Exception as e:
                # Skip invalid questions
                print(f"Error parsing question: {str(e)}")
                continue

        if not questions:
            raise HTTPException(
                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                detail="Failed to parse generated questions. Please try again."
            )

        return GenerateQuestionsResponse(
            questions=questions,
            count=len(questions),
            source="gemini"
        )

    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Error generating questions: {str(e)}"
        )


@router.post("/generate/study-plan")
async def generate_study_plan(
    class_level: int,
    weak_topics: List[str],
    target_exam: str,
    days_available: int,
    current_user: dict = Depends(get_current_user)
):
    """
    Generate a personalized study plan using Gemini AI

    Args:
        class_level: CBSE class (10 or 12)
        weak_topics: List of topics to focus on
        target_exam: Target exam (Board Exams, JEE, NEET, etc.)
        days_available: Number of days until exam
    """
    try:
        if class_level not in [10, 12]:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Class level must be 10 or 12"
            )

        if days_available < 1:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Days available must be at least 1"
            )

        # Generate study plan using Gemini
        study_plan = await gemini_client.generate_study_plan(
            class_level=class_level,
            weak_topics=weak_topics,
            target_exam=target_exam,
            days_available=days_available
        )

        return study_plan

    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Error generating study plan: {str(e)}"
        )
